lib/rack/attack/throttle_with_leaky_bucket.rb

```diff
@@ -0,0 +1,68 @@
+module Rack
+  class Attack
+    class ThrottleWithLeakyBucket
+      MANDATORY_OPTIONS = [:capacity, :leak]
+      attr_reader :name, :capacity, :leak, :block, :type
+
+      def initialize(name, options, block)
+        @name, @block = name, block
+        MANDATORY_OPTIONS.each do |opt|
+          raise ArgumentError.new("Must pass #{opt.inspect} option") unless options[opt]
+        end
+        @capacity = options[:capacity]
+        @leak     = options[:leak]
+        @type     = options.fetch(:type, :throttle_with_leaky_bucket)
+      end
+
+      def cache
+        Rack::Attack.cache
+      end
+
+      def [](req)
+        discriminator = block[req]
+        return false unless discriminator
+
+        # Normalize blocks to values
+        current_capacity = normalize_block(capacity, req)
+        current_leak     = normalize_block(leak, req)
+
+        # Read the bucket data and unserialize it. We only update the bucket data
+        # if we've changed the value. We don't write to update the leaked amount
+        # since that can be calculated and since the TTL will remove the item when
+        # it has drained.
+        key = "#{name}:#{discriminator}"
+        bucket = LeakyBucket.unserialize(cache.read(key), current_capacity, current_leak)
+        throttled = bucket.full?
+        bucket.add(1) unless bucket.full?
+        store_bucket(key, bucket) if bucket.updated?
```

Comment:
@jeroenvisser101 Since the cached value is read in line 34, and written in line 37, this implementation is not atomic, and can lead to uncounted requests.